### PR TITLE
WT-2318 Add implementation for auto adjusting condition variables.

### PR DESCRIPTION
**Pull request by agorrod**
Where the amount of time spent waiting varies depending on whether work is being done between timeouts.

### CODE REVIEW

**Comment by agorrod** [comment]
@sueloverso I've written up the wrapper functions we talked about this morning. Would you mind taking a look at switching some of the condition wait usage in the logging code to this new mechanism?

It might be interesting to track some statistics about how often different code paths are being taken as well.

File: src/conn/conn_log.c [hunk 1]
```diff
@@ -793,8 +817,8 @@ __log_server(void *arg)
 		}
 
 		/* Wait until the next event. */
-		WT_ERR(__wt_cond_wait_signal(session, conn->log_cond,
-		    WT_MILLION / WT_FORCE_PER_SECOND, &signalled));
+		WT_ERR(__wt_cond_auto_wait_signal(session, conn->log_cond,
+		    did_work || signalled, &signalled));
```
Comment (agorrod) [inline src/conn/conn_log.c]: I don't think you need signalled in the condition - the auto conditionals clear the timer automatically when woken by a signal.

**Comment by agorrod** [comment]
autobuild: retest this please

**Comment by sueloverso** [comment]
@agorrod I wrote a test program to look for high % CPU usage.  The only problem is that the value is quite variable across different machines.  I set the max to 10.0 percent after some experiementation.  On my Mac it was 0.0, on my AWS box, 8.3 and on the pull-request-tester machine it was 3.0.  But 10.0 is lower than the current-develop value when I run it on my Mac (and as high as 42% on my AWS box!!).  So the test fails everywhere on develop and passes everywhere, for now, with this branch.  We can revisit if the turns out to be problematic.

**Comment by agorrod** [comment]
Thanks @sueloverso I'm happy for this to be merged.

**Pull request merged**